hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/async/RouterAsyncRpcClient.java

```diff
@@ -479,6 +488,76 @@ public <T extends RemoteLocationContext, R> Map<T, R> invokeConcurrent(
     return asyncReturn(Map.class);
   }
 
+  @SuppressWarnings("unchecked")
+  public <T extends RemoteLocationContext, R> List<RemoteResult<T, R>> invokeConcurrent(
+      final Collection<T> locations, final RemoteMethod method,
+      boolean standby, long timeOutMs,
+      Class<R> clazz) throws IOException {
+
+    final UserGroupInformation ugi = RouterRpcServer.getRemoteUser();
+    final Method m = method.getMethod();
+
+    if (locations.isEmpty()) {
+      throw new IOException("No remote locations available");
+    } else if (locations.size() == 1 && timeOutMs <= 0) {
+      // Shortcut, just one call
+      return invokeSingle(locations.iterator().next(), method);
+    }
+    // Don't acquire CONCURRENT_NS permit here.
+    RouterRpcFairnessPolicyController controller = getRouterRpcFairnessPolicyController();
+
```

Comment:
I understand that you don't want to acquirePermit(CONCURRENT_NS, ugi, method, controller) here. You can do special processing for the case where ns=CONCURRENT_NS in RouterAsyncRpcClient#acquirePermit  or RouterAsyncRpcFairnessPolicyController without duplicating this method again.

Reply:
Do extra logic in RouterAsyncRpcFairnessPolicyController#acquirePermit.  when nsid equals CONCURRENT_NS, return true and in RouterAsyncRpcFairnessPolicyController#releasePermit return immediately when nsid equals CONCURRENT_NS.